examples/tinyhttp.rs

```diff
@@ -169,9 +168,13 @@ impl Decoder for Http {
             };
 
             let toslice = |a: &[u8]| {
-                let start = a.as_ptr() as usize - src.as_ptr() as usize;
-                assert!(start < src.len());
-                (start, start + a.len())
+                if (a.as_ptr() as usize) > (src.as_ptr() as usize) {
+                    let start = a.as_ptr() as usize - src.as_ptr() as usize;
+                    assert!(start < src.len());
+                    (start, start + a.len())
+                } else {
+                    (0, a.len())
+                }
```

Comment:
This change cannot possibly be correct. If `a.as_ptr() < src.as_ptr()`, then there's something weird going on. How did we end up with a slice that's not a sub-slice of `src`?

Reply:
The fact is that I met it at local, do not get the root cause :(

Reply:
I tried to use wrk to test it's performance, but met this problem also.
Then i found the httparse is impure.

@Darksonn  

https://docs.rs/httparse/latest/src/httparse/lib.rs.html#479
```rs
        const GET: [u8; 4] = *b"GET ";
        const POST: [u8; 4] = *b"POST";
        let method = match bytes.peek_n::<[u8; 4]>(4) {
            Some(GET) => {
                unsafe {
                    bytes.advance_and_commit(4);
                }
                "GET"
            }
            Some(POST) if bytes.peek_ahead(4) == Some(b' ') => {
                unsafe {
                    bytes.advance_and_commit(5);
                }
                "POST"
            }
            _ => complete!(parse_token(&mut bytes)),
        };
        self.method = Some(method);
```

Reply:
@biluohc nice finding!

Reply:
Okay, so the problem is that it sometimes returns references into static memory. Currently, you just pretend that if this happens, then actually the string contains the `len` first characters of the buffer, no matter what the static string actually contains. This happens to work for GET/POST since that is actually the first 3 or 4 bytes in the buffer, but that's a serious coincidence and certainly not something we can rely on.

Reply:
From its implementation, we can see it only handle GET/POST in this way.

Reply:
Sorry, but this is way too fragile to future changes in httparse. We have already seen them change this kind of thing once (that's what caused this issue), so they could easily make a similar change to something else in the future.

Reply:
Hello, I have an idea, 
how about getting the Method first, so that don’t need to access its slice later.

https://docs.rs/http/latest/http/method/struct.Method.html#method.from_bytes

Code might as follows, it maybe robust enough:
```rs
 let (method, path, version, amt) = {
// ..
            (
                Method::from_bytes(r.method.unwrap().as_bytes()).map_err()?,
                toslice(r.path.unwrap().as_bytes()),
                r.version.unwrap(),
                amt,
            )
};
```

Reply:
It still assumes that lib handles method in this way. Since it is an example, not for production, could I just use string copy here?

Reply:
ping @Darksonn

Reply:
I use windows to find the index instead of the hardcode.